internal/healthchecks/generic/generic.go

```diff
@@ -0,0 +1,198 @@
+			replicaSet := obj.(*appsv1.ReplicaSet)
+			if err := client.Get(ctx, types.NamespacedName{
+				Name:      replicaSet.GetName(),
+				Namespace: replicaSet.GetNamespace(),
+			}, replicaSet); err != nil {
+				gvkErrorsMap[toErrKey(replicaSet)] = err.Error()
+				continue
+			}
+			if replicaSet.Status.AvailableReplicas != replicaSet.Status.Replicas {
+				gvkErrorsMap[toErrKey(replicaSet)] = "ReplicaSet is not ready"
+			}
+		case *corev1.Pod:
+			pod := obj.(*corev1.Pod)
+			if err := client.Get(ctx, types.NamespacedName{
+				Name:      pod.GetName(),
+				Namespace: pod.GetNamespace(),
+			}, pod); err != nil {
+				gvkErrorsMap[toErrKey(pod)] = err.Error()
+				continue
+			}
+			if pod.Status.Conditions == nil {
+				gvkErrorsMap[toErrKey(pod)] = "Pod has no conditions"
+				continue
+			}
+			if pod.Status.Phase != corev1.PodRunning && pod.Status.Phase != corev1.PodSucceeded {
+				gvkErrorsMap[toErrKey(pod)] = pod.Status.Message
+			}
+		case *apiregistrationv1.APIService:
+			apiService := obj.(*apiregistrationv1.APIService)
+			if err := client.Get(ctx, types.NamespacedName{
+				Name:      apiService.GetName(),
+				Namespace: apiService.GetNamespace(),
+			}, apiService); err != nil {
+				gvkErrorsMap[toErrKey(apiService)] = err.Error()
+				continue
+			}
+			if apiService.Status.Conditions == nil {
+				gvkErrorsMap[toErrKey(apiService)] = "APIService has no conditions"
+				continue
+			}
+			for _, condition := range apiService.Status.Conditions {
+				if condition.Type == apiregistrationv1.Available && condition.Status != "True" {
+					gvkErrorsMap[toErrKey(apiService)] = condition.Message
+				}
+			}
+		case *apiextensionsv1.CustomResourceDefinition:
+			crd := obj.(*apiextensionsv1.CustomResourceDefinition)
+			if err := client.Get(ctx, types.NamespacedName{
+				Name:      crd.GetName(),
+				Namespace: crd.GetNamespace(),
+			}, crd); err != nil {
+				gvkErrorsMap[toErrKey(crd)] = err.Error()
+				continue
+			}
+			if crd.Status.Conditions == nil {
+				gvkErrorsMap[toErrKey(crd)] = "CustomResourceDefinition has no conditions"
+				continue
+			}
+			for _, condition := range crd.Status.Conditions {
+				if condition.Type == apiextensionsv1.Established && condition.Status != "True" {
+					gvkErrorsMap[toErrKey(crd)] = condition.Message
+				}
+			}
+		default:
+			// If we don't know how to check the health of the object, we assume it's healthy.
+			continue
+		}
+	}
+
+	if len(gvkErrorsMap) == 0 {
+		return true, nil
+	}
+	return false, gvkErrorsMapToErr(gvkErrorsMap)
+}
+
+// toErrKey returns a string that uniquely identifies a resource based on its GVK and name.
+// This is used to create a map of "GVK + Resource Name" to error message.
+func toErrKey(resource client.Object) string {
+	return fmt.Sprintf("%s.%s.%s/%s", resource.GetObjectKind().GroupVersionKind().Kind, resource.GetObjectKind().GroupVersionKind().Group, resource.GetObjectKind().GroupVersionKind().Version, resource.GetName())
```

Comment:
This key needs to include a namespace to cover the case where a manifest has two or more objects with the same GVK and name, but in different namespaces.

Reply:
done